script_test.go

```diff
@@ -1768,6 +1768,44 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
 	}
 }
 
+func TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice(t *testing.T) {
+	t.Parallel()
+	os.Setenv("ENV1", "test1")
```

Comment:
Not a bad idea in principle, but unfortunately `Setenv` isn't safe to use in a parallel test, because it affects the environment for all tests. (If we _were_ going to do that, we could use `t.Setenv`, which does the cleanup automatically, but it's still not parallelisable.)

Actually, what we care about is not the whole _test binary's_ environment, but just the _pipe's_, isn't it? When you frame it that way, all we need to do is create a pipe with some environment variable already set, and check that `WithEnv` unsets it. Something like this, perhaps?

```go
func TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice(t *testing.T) {
	t.Parallel()
	p := script.NewPipe().WithEnv([]string{"ENV1=test1"}).Exec("sh -c 'echo ENV1=$ENV1'")
	want := "ENV1=test1\n"
	got, err := p.String()
	if err != nil {
		t.Fatal(err)
	}
	if got != want {
		t.Fatalf("want %q, got %q", want, got)
	}
	got, err = p.Echo("").WithEnv([]string{}).Exec("sh -c 'echo ENV1=$ENV1'").String()
	if err != nil {
		t.Fatal(err)
	}
	want = "ENV1=\n"
	if got != want {
		t.Errorf("want %q, got %q", want, got)
	}
}
```

Reply:
Great idea! Thanks @bitfield